### ISSUE TITLE
WordPress pagination links revert to shop pagination links in some cases

### ISSUE THREAD

**Issue by michaelbeil** [AffiliateWP/AffiliateWP#1355]
It appears like we need to account for another instance where WooCommerce product pagination is breaking when filtering by attributes. 

Here is what happens: 
- Go to http://yoursite.dev/shop/products/?filter_material=66
- When a user navigates to page 2, then http://yoursite.dev/shop/products/page/2/ is displayed instead of http://yoursite.dev/shop/products/page/2/?filter_material=66  

Related: #955, #1277, #1290, #1291
 
**(Update):**
This is affecting 2 WooCommerce sites ([#1](https://secure.helpscout.net/conversation/215405333/36276/?folderId=634609), [#2](https://secure.helpscout.net/conversation/213080668/35956/?folderId=634609)) and [1 bbPress site](https://secure.helpscout.net/conversation/215906011/36328/?folderId=634609).

There are three reports of this breaking blog pagination. 

I think we should come up with a less fragile solution to “rewriting” the URLs, and instead create them from scratch in certain situations.

**Comment by pippinsplugins** [comment]
@michaelbeil does it only happen when a referral variable is present?

**Comment by michaelbeil** [comment]
No, it happens regardless of a referral variable.

**Comment by michaelbeil** [comment]
My thoughts are to add meta_is_product_attribute to line 557 in includes/integrations/class-woocommerce.php: 

if ( ( ! is_shop() && ! is_paged() ) || is_product_taxonomy() || is_search() || meta_is_product_attribute() ) {

The issue is that other plugins and WC extensions might use these arguments to provide some information that still needs to be included in the pagination. 

Should we simply remove the referral variable  (from the URL, still storing the cookie) if an affiliate link is used instead of removing everything except for on certain conditions?

**Comment by pippinsplugins** [comment]
Does adding `meta_is_product_attribute()` to the conditional resolve it?

**Comment by michaelbeil** [comment]
I have not been able to replicate it locally, but only on two customer sites so far. I have not received admin access, so I am still attempting to figure out how they are filtering by attribute.

**Comment by nhadsall** [comment]
We were effected by this on one of our sites. The bug is that aff wp strips ALL query strings from the URL when it attempts to strip the referral code. The approach currently used assumes that no other plugin will modify information after the pagination URL (appending query strings). In this case it is WooCommerce itself that adds the query string.

We do not use pretty URLs, so it is a simple matter of just changing the function to:
`$link = remove_query_arg(affiliate_wp()->tracking->get_referral_var(), $link);
return $link;`

Perhaps you could do this, plus change the regex to get everything before and after the pretty URL (ie. preserving any query string).

**Comment by DrewAPicture** [comment]
Added a PR: #1360

**Comment by michaelbeil** [comment]
Testing.

**Comment by michaelbeil** [comment]
This issue will be used to deprecate the `strip_referral_from_paged_urls` method in the WC integration class so we can address this plugin-wide.

**Comment by DrewAPicture** [comment]
Now that we're in the global plugin scope, I've added a couple of tests to the PR.

**Comment by michaelbeil** [comment]
Everything is working well here.

**Comment by pippinsplugins** [comment]
Hmm, this isn't working for me.

The referral parameters are still included on the paginated links.

**Comment by pippinsplugins** [comment]
Oh wait, yes it is.

This is interesting though and I wonder if we should account for it.

Let's say a URL like this is used:
```
http://local.dev/woo/shop/?ref=jane
```
Clicking onto page 2 keeps the parameter in the URL, resulting in this:
```
http://local.dev/woo/shop/page/2/?ref=jane
```

If, however, we start at a paged URL, the next page URL has the parameter removed. `http://local.dev/woo/shop/page/2/?ref=jane` becomes `http://local.dev/woo/shop/page/3`.

I'm inclined to think we should make it consistent.  I'm also fine ignoring it as it does not appear to cause any problems.

**Comment by michaelbeil** [comment]
This is specifically with pretty affiliate links. What happens when you use a pretty affiliate link on a shop page? 

Although, yea it should probably be consistent with non-pretty affiliate links as well when there is a traveling referral variable.

**Comment by pippinsplugins** [comment]
It happens in exactly the same way with both pretty and non-pretty links.

**Comment by michaelbeil** [comment]
Oh, I see. Maybe @sumobi can test too and let us know what he sees. I see the parameter persists with non-pretty affiliate links, but not pretty affiliate links. Though, this round of tests caused the referral variable to be prepended prior page/2. Not sure why that is happening.

**Comment by DrewAPicture** [comment]
@pippinsplugins:
> Clicking onto page 2 keeps the parameter in the URL, resulting in this:
>
> http://local.dev/woo/shop/page/2/?ref=jane

I can't reproduce this. The initial page load will include the referral var and *any* subsequent paged URL should get stripped.

@michaelbeil:
> Though, this round of tests caused the referral variable to be prepended prior to page/2. Not sure why that is happening.

Also can't reproduce that either, though WordPress simply appends paging endpoints to the end of the existing URL (when this code isn't applied, that is).

**Comment by pippinsplugins** [comment]
Definitely happens to me.

I load the shop page with a ref and then look at the pagination URLs:
![screen shot 2016-06-21 at 8 53 24 am](https://cloud.githubusercontent.com/assets/1034109/16231854/af5eeba4-378d-11e6-8319-cd60fcfd504a.png)

**Comment by DrewAPicture** [comment]
Hmm, that shouldn't be happening (obviously).

**Comment by DrewAPicture** [comment]
Posted my screenshot too. I wonder what could possible cause that ... hmm.

**Comment by DrewAPicture** [comment]
Removed the `is_paged()` check.

**Comment by michaelbeil** [comment]
Removing the `is_paged()` check has resolved the traveling referral variable for both pretty and non-pretty affiliate links. Blog pagination, shop pagination, search pagination, and category pagination are all working as expected.

**Comment by DrewAPicture** [comment]
Tests passing. Merged #1360.
